### PR TITLE
Implement a composition graph library.

### PR DESCRIPTION
**Pull request by peterhuene**
This PR implements a composition graph library that will serve as the basis for how WAC does AST resolution in the near future.

The inspiration comes from the existing `CompositionGraph` API in `wasm-compose`, except this API is more flexible, allowing a composition graph to define its own types, have explicit imports and aliases, and export any node in the graph.

The type representation present in `wac-parser` has been extracted to its own crate, `wac-types`. It will be shared in the future between the parser and the graph library.

As a result, a chunk of this new code is just copied out of `wac-parser`. Once the graph library is merged, a follow-up commit will eliminate the duplicated code and change `wac-parser` to resolve ASTs to a `CompositionGraph`.

### CODE REVIEW

**Comment by peterhuene** [comment]
Note that the new `crates/wac-types/src/component.rs`, `crates/wac-types/src/core.rs`, and `crates/wac-types/src/package.rs` files are mostly a direct copy of what's in the `wac-parser` crate.

Once this merges, I plan on updating `wac-parser` and eliminating a huge chunk of the duplicated code.

I'll be pushing up more test cases shortly.

File: crates/wac-graph/src/graph.rs
Comment (rylev) [inline crates/wac-graph/src/graph.rs:807]: The error here and in `export_node` is incorrect (these aren't package ids).

File: crates/wac-graph/src/graph.rs
Comment (rylev) [inline crates/wac-graph/src/graph.rs:806]:
```suggestion
    pub fn set_node_name(&mut self, node: NodeId, name: impl Into<String>) -> GraphResult<()> {
```

File: crates/wac-graph/src/graph.rs
Comment (rylev) [inline crates/wac-graph/src/graph.rs:240]: Just for my own education: are you using generations here so you don't have to reallocate the entire graph if you remove a node?

File: crates/wac-graph/src/graph.rs
Comment (rylev) [inline crates/wac-graph/src/graph.rs:208]: I worry that "type aggregation" is going to be meaningless to almost all users. Is there some way we can phrase this that's a bit more meaningful?

**Review by rylev — APPROVED** [review]
Awesome! I've already given the API a try [here](https://github.com/rylev/wac-example/blob/main/programmatic/src/main.rs), and overall, while it's not the absolute easiest API to use, I think it's a good basis to build on top of.

Most of my feedback was on naming and error messages.

File: crates/wac-graph/src/encoding.rs
Comment (peterhuene) [reply crates/wac-graph/src/encoding.rs:149]: A pop should always be paired with a push, so I didn't make it optional; it should panic instead as it is 100% a bug if the stack is empty.

File: crates/wac-graph/src/graph.rs
Comment (peterhuene) [reply crates/wac-graph/src/graph.rs:1777]: I think that makes sense to have, I'll see about adding it.

File: crates/wac-graph/tests/encoding.rs
Comment (peterhuene) [reply crates/wac-graph/tests/encoding.rs:63]: Can do.

File: crates/wac-graph/src/graph.rs
Comment (peterhuene) [reply crates/wac-graph/src/graph.rs:672]: Yeah, the graph API is closer to the component model primitives than WAC; although it's called `wac-graph` as a crate, it has little to do with WAC itself other than being the primitive upon which AST resolution and encoding will soon rest.

File: crates/wac-graph/src/graph.rs
Comment (peterhuene) [reply crates/wac-graph/src/graph.rs:301]: Once WAC AST resolution is using the composition graph, the resolver will store context per-node relating back to AST nodes (spans, other information, etc.).

If you feel like this unnecessarily complicates the graph, I'm totally fine with removing it and having the AST resolver keep its own mapping from node to contextual data.

File: crates/wac-graph/tests/graphs/implicit-import-conflict/error.txt
Comment (peterhuene) [reply crates/wac-graph/tests/graphs/implicit-import-conflict/error.txt:4]: Definitely should have the package name in there, that's easy to do.

File: crates/wac-graph/src/graph.rs
Comment (peterhuene) [reply crates/wac-graph/src/graph.rs:807]: Whoops, good catch!

File: crates/wac-graph/src/graph.rs
Comment (peterhuene) [reply crates/wac-graph/src/graph.rs:208]: Agreed, it should be meaningless; let me see if I can word this in a way that makes sense to users.

File: crates/wac-graph/src/graph.rs
Comment (peterhuene) [reply crates/wac-graph/src/graph.rs:301]: I'm now actually of the opinion that it complicates an already complicated API surface; I'll remove the context storage from the graph and users can just store contextual data themselves.

Comment (rylev) [reply crates/wac-graph/src/graph.rs:301]: If we can avoid having all consumers of the graph have to think about this extra context param, I think that's a usability win, but I'll let you decide if that makes the WAC AST to graph mapping much harder.

File: crates/wac-graph/src/graph.rs
Comment (peterhuene) [reply crates/wac-graph/src/graph.rs]: I ended up naming this `alias_instance_export` and renamed a few of the other "add node" methods to hopefully make a little more sense.

File: crates/wac-graph/src/graph.rs
Comment (peterhuene) [reply crates/wac-graph/src/graph.rs]: Removed the `C` parameter.

File: crates/wac-graph/src/graph.rs
Comment (peterhuene) [reply crates/wac-graph/src/graph.rs]: I renamed it to `connect_argument` and `disconnect_argument`; I'm not sure if that's better or worse, though.

File: crates/wac-graph/src/graph.rs
Comment (peterhuene) [reply crates/wac-graph/src/graph.rs]: So it's generally not a single function call to add a type to the `Types` collection as types are fairly nested within their definitions (i.e. lots of internal ids stored).

I've left it as-is for now and we can maybe improve this in the future.

File: crates/wac-graph/src/graph.rs
Comment (rylev) [reply crates/wac-graph/src/graph.rs]: This might be an example of where I'm thinking about things less like a graph than I should be, but the high order bit for me is answering question "an argument to *what*?". Mentioning "connection" or "edge" isn't as necessary to me as the graphness of it all isn't at the forefront of my mind. That's why `instantiation_argument` makes more sense to me as it answers the question I was interested in. 

That all being said, I can imagine with more advanced use cases that having the graph like terminology at the forefront might make things more clear, so I'm not necessarily pushing one way or the other, but hopefully it's clear now why I prefer `instantiation_argument`.

File: crates/wac-graph/src/graph.rs
Comment (rylev) [reply crates/wac-graph/src/graph.rs]: @peterhuene thoughts on this?

> I would also reorder the arguments here. I'm not sure what the best order is, but I would keep argument (nee source) and argument_name (nee argument) together.

File: crates/wac-graph/src/graph.rs
Comment (peterhuene) [reply crates/wac-graph/src/graph.rs]: I generally like to have a verb in the method name, like `alias_instance_export` has "alias" as the verb.

Perhaps just `set_instantiation_argument`?

Comment (peterhuene) [reply crates/wac-graph/src/graph.rs]: I went with `set_instantiation_argument`/`unset_instantiation_argument`; with the rearranged parameters, I think it's an improvement over `connect_argument`/`disconnect_argument`.

File: crates/wac-graph/src/graph.rs
Comment (peterhuene) [reply crates/wac-graph/src/graph.rs]: The order is now "instantiation, argument_name, argument", as it's "for this instantiation and argument name, set the argument".

**Review by fibonacci1729 — APPROVED** [review]

**Review by rylev — APPROVED** [review]

**Pull request merged**
